scrapy_autoextract/middlewares.py

```diff
@@ -145,12 +145,24 @@ def process_response(self, request, response, spider):
         url = request.meta[AUTOEXTRACT_META_KEY]['original_url']
 
         try:
-            result = json.loads(response.body.decode('utf8'))[0]
+            response_object = json.loads(response.body.decode('utf8'))
         except Exception:
             self.inc_metric('autoextract/errors/json_decode')
             raise AutoExtractError('Cannot parse JSON response from AutoExtract'
                                    ' for {}: {}'.format(url, response.body))
 
+        result = None
+        if isinstance(response_object, list):
+            result = response_object[0]
+        elif isinstance(response_object, dict):
+            self.inc_metric('autoextract/errors/result_error')
+            raise AutoExtractError('Received error from AutoExtract for '
```

Comment:
We should be raising errors based on http status code, not based on list/dict in the response.

Reply:
Writing the code for this is really easy, but I'm not sure what's the best logic to handle this.

`response.status != 200` is the best way to handle the errors, I agree.

But if the response is OK, we just return whatever object we receive from the API? The API should normally return a list in case of status=200 because that's a successful request.
Is it necessary to check the type and handle that as well in the middleware?

Reply:
> response.status != 200 is the best way to handle the errors, I agree.

We need to be careful, and make sure that e.g. 429 errors are retried, not handled as usual.  It may depend on middleware order.

> But if the response is OK, we just return whatever object we receive from the API? The API should normally return a list in case of status=200 because that's a successful request.

A good point. I think it makes sense to return a first element. If we add options to customize request, and add e.g. batching support, we may change this.

> Is it necessary to check the type and handle that as well in the middleware?

I thinkn no, currently we don't need to check response type. We expect list, if that's not a list, something is not working as intended, and so exception (even IndexError) is fine.

Reply:
I pushed commit https://github.com/scrapinghub/scrapy-autoextract/pull/2/commits/d835936ca5d4760778160951c691d5829761f60a ; I know this doesn't solve everything, because it doesn't handle HTTP 429.

In that case we have a few options to retry:
* derive AutoExtractMiddleware from RetryMiddleware and call _retry in case of 429
* add another middleware maybe called AutoExtractRetryMiddleware and derive it from RetryMiddleware and call _retry there, so people could use that if they want, add it in the correct priority order
* other options?

Reply:
If Scrapy doesn't handle 429 properly, we need to fix Scrapy, and probably provide a backport of the fix as a middleware here. If it works properly already, it is a matter of making sure the order of middlewares is right.

Do you know if 429 error is caught here or not, if user follows the guideline in README?

Reply:
In a spider settings I added:

```py
RETRY_ENABLED = True
RETRY_HTTP_CODES = [429]
RETRY_TIMES = 3
```

I hacked `response.status = 429` in process_response function and I tested like crazy.
In Scrapy, the default RetryMiddleware has order 550 in downloader middlewares.

Scrapy doesn't retry the failed request, but I believe it's because we raise AutoExtractError if response.status != 200. Maybe in this case we should return the request.

I'll check more.

Reply:
Correction. If I add code 200 to the RETRY_HTTP_CODES, the retry works. The RetryMiddleware doesn't see the hacked 429 code from the AutoExtractMiddleware, so that was the issue.

I'll check more.

Reply:
In conclusion, Scrapy Retry handles 429 correctly, even when we raise the AutoExtract error.
The only thing missing is to delay the next request, to slow down.

Reply:
How I checked that Scrapy Retry middleware handles HTTP 429: I created a mock server that simulates error responses coming from the real AutoExtract API server.
I pointed AutoExtract Spiders to use this branch of the middleware and defined: `AUTOEXTRACT_URL = 'http://127.0.0.1:8080/v1/extract'`

```
INFO: Using AutoExtract API URL: http://127.0.0.1:8080/v1/extract
DEBUG: Process AutoExtract request for article URL <AutoExtract http://quotes.toscrape.com/>
INFO: Retrying <AutoExtract http://127.0.0.1:8080/v1/extract> (failed 1 times): 429 Unknown Status
INFO: Retrying <AutoExtract http://127.0.0.1:8080/v1/extract> (failed 2 times): 429 Unknown Status
Gave up retrying <AutoExtract http://127.0.0.1:8080/v1/extract> (failed 3 times): 429 Unknown Status
2019-12-09 14:08:10 [scrapy_autoextract.middlewares] DEBUG: AutoExtract response status=429  headers={'content-type': 'application/problem+json', 'content-language': 'en', 'www-authenticate': 'Basic realm="User Visible Realm"', 'date': 'Mon, 09 Dec 2019 14:08:10 GMT', 'server': 'Python/3.6 aiohttp/3.6.2'}  content={"type": "System-wide rate limit exceeded", "title": "http://errors.xod.scrapinghub.com/rate-limit-exceeded.html"}
```

*Side note*: Aparently Twisted doesn't recognize HTTP 429, it's not defined in https://github.com/twisted/twisted/blob/trunk/src/twisted/web/_responses.py so that's why the error from Retry appears as "429 Unknown Status".